### PR TITLE
UPE: migrate the upe javascript to use the wp-scripts pipeline (webpack)

### LINKED ISSUE #1661: SPIKE - Stripe+UPE - Webpack/Assets compilation
## Description

We will be fully porting UPE and the new settings design to the Stripe plugin: pc4etw-hj-p2
It sounds like some work has already started to include Webpack in the plugin.

## Questions
- What is remaining to do?
- What approach are we going to follow for the tests?
- What linting rules do we need?

### PR DESCRIPTION
**Pull request by diegocurbelo**
# Changes proposed in this Pull Request:

Issue: 
Fixes #

This PR migrate the UPE javascript for classic checkout from normal js asset + minified, to use the wp-scripts pipeline (webpack)

# Testing instructions
- Enable UPE feature flag: 
```sql
INSERT INTO `wp_options` (`option_id`, `option_name`, `option_value`, `autoload`) VALUES (NULL, '_wcstripe_feature_upe', '1', 'no') ON DUPLICATE KEY UPDATE `option_value`=VALUES(`option_value`);
```
- Run `npm install`
- Run `npm run start:webpack`
- Check that the folder `/build` contains the following files: `style-upe_classic.css`, `upe_classic.js` and `upe_classic.asset.php`

- Go to WooCommerce > Settings > Payments > Stripe settings
- Check the "Enable new checkout experience" option and hit save
- Scroll to the bottom and there should now be a table of UPE payment methods, make sure Card is turned on.

- Go to `http://localhost:8082/checkout/`, and check that the Credit Card form loads correctly.

### CODE REVIEW

File: client/classic/api.js
Comment (reykjalin) [inline client/classic/api.js]: Is there a reason we didn't add this to the `client/api` folder? 🤔 

I chose not to adapt the full WCPay API interaction code when I did the Blocks refactoring just to keep things simple, but even if we want to do that I still think this should live in `client/api` 🙂

File: client/classic/upe.js
Comment (reykjalin) [inline client/classic/upe.js]: We can also `import $ from jQuery` if we prefer to use that syntax here. You can look at `client/api/index.js` for an example of that 🙂

File: client/classic/api.js
Comment (asumaran) [inline client/classic/api.js]: Is this method used anywhere, or is it going to be? I couldn't find a place where this it's used.

File: client/utils/index.js
Comment (asumaran) [inline client/utils/index.js]: If `ajaxURL` param is a `string` why do we call `.toString()` here? Just to make sure?

File: client/utils/index.js
Comment (asumaran) [inline client/utils/index.js]: typo?

>The value of the parameter _**or**_ null.

File: includes/payment-methods/class-wc-stripe-upe-payment-gateway.php
Comment (asumaran) [inline includes/payment-methods/class-wc-stripe-upe-payment-gateway.php]: Any specific reason why this code block was moved a few lines above?

File: includes/payment-methods/class-wc-stripe-upe-payment-gateway.php
Comment (asumaran) [inline includes/payment-methods/class-wc-stripe-upe-payment-gateway.php:226]: Is this equivalent to "Statement Descriptor" we have on Stripe?

Screen Shot 2021-08-02 at 7 33 47 PM

**Review by asumaran — COMMENTED** [review]
I assume this is a WIP PR that belongs to a more extensive PR? I left a few comments nevertheless.

File: client/classic/api.js
Comment (diegocurbelo) [reply client/classic/api.js]: Good point... I didn't want to move the `stripe-utils` folder to the same level as `blocks`, `classic` and `api`.. but you are right that we should have only 1 API file, I'll change it

File: client/classic/upe.js
Comment (diegocurbelo) [reply client/classic/upe.js]: I like the `import $` but I'm not sure how to implement the early return when we don't have stripe keys:
```
if ( ! publishableKey ) {
	// If no configuration is present, probably this is not the checkout page.
	return;
}
```

Comment (reykjalin) [reply client/classic/upe.js]: Yeah, that's a good point 🤔 
I'm sure it's possible, but it's probably going to be way worse than using what we already have 😅 

I think we could maybe `import jQuery` (or `import jQuery from jQuery`?) to be explicit about the dependency, but not change anything in the code? 🤔

Comment (diegocurbelo) [reply client/classic/upe.js]: Added the explicit import 👍🏼

File: client/utils/index.js
Comment (diegocurbelo) [reply client/utils/index.js]: Fixed

File: client/utils/index.js
Comment (diegocurbelo) [reply client/utils/index.js]: When merging the API files, this method was no longer needed.

File: includes/payment-methods/class-wc-stripe-upe-payment-gateway.php
Comment (diegocurbelo) [reply includes/payment-methods/class-wc-stripe-upe-payment-gateway.php]: Just as a precaution, to make sure `payment_methods` are loaded before calling any auxiliary methods (just in case they need them)

```
...

// Load the form fields.
$this->init_form_fields();

// Load the settings.
$this->init_settings();
```

File: includes/payment-methods/class-wc-stripe-upe-payment-gateway.php
Comment (diegocurbelo) [reply includes/payment-methods/class-wc-stripe-upe-payment-gateway.php:226]: It's a TODO because in this PR the gateway is an independent one (called Stripe UPE), but with the changes in the admin settings, after merging the PRs it will need to be changed to be a config of the parent class (or deleted if not it's not necessary)

File: client/stripe-utils/utils.js
Comment (reykjalin) [inline client/stripe-utils/utils.js]: This will always be `true` if `wc.wcSettings.getSetting( 'stripe_data', null )` returns `null` because `undefined !== null` evaluates to `true` when we want it to evaluate to `false` 🙂 

Let's change to `wc.wcSettings.getSetting( 'stripe_data', undefined )`.

File: client/stripe-utils/utils.js
Comment (reykjalin) [inline client/stripe-utils/utils.js:4]: I'm pretty against relying on global values; it makes the code much more difficult to understand, especially when you don't know where the global comes from — or if you miss the `/* global some_variable */` comment.

Let's remove the `wc` global and go back to using the imported `getSetting` instead.

I'd personally prefer to see us not rely on globals in any way, and instead assign `wc_stripe_upe_params` to the `wc.wcSettings` object so we can use `getSetting( 'stripe_upe' )` or something along those lines instead.

File: includes/class-wc-stripe-intent-controller.php
Comment (reykjalin) [inline includes/class-wc-stripe-intent-controller.php:269]: Is there a reason to revert to a generic `Exception`? I think in general it would be beneficial to keep the more specific `Process_Payment_Exception` here to give more context 🤔

File: includes/payment-methods/class-wc-stripe-upe-payment-gateway.php
Comment (reykjalin) [inline includes/payment-methods/class-wc-stripe-upe-payment-gateway.php:205]: I wonder if we can't just inherit the `javascript_params()` function from the parent class and use `$this->javascript_params()` directly?

File: includes/payment-methods/class-wc-stripe-upe-payment-gateway.php
Comment (reykjalin) [inline includes/payment-methods/class-wc-stripe-upe-payment-gateway.php]: Is there a reason we changed this here instead of updating the JS?

To be clear; I'm fine with this change, I just wonder if it wouldn't be better to re-use the parent JS function and use the same configuration object throughout the whole JS codebase instead of having different names for each configuration object depending on whether you're working in the Blocks, Shortcode, or UPE JS 🤔

**Review by reykjalin — CHANGES_REQUESTED** [review]
Sorry, I unfortunately don't have time to test this today, but I did see some things in the code that I think we need to discuss some more.

I'll test this on Monday.

File: client/stripe-utils/utils.js
Comment (diegocurbelo) [reply client/stripe-utils/utils.js]: I'm not sure I understand the change proposed... the ternary if is checking if `wc_stripe_upe_params` is defined (`'undefined' !== typeof wc_stripe_upe_params`), and if it is, it will be assigned to `stripeServerData`, and if not `wc.wcSettings.getSetting( 'stripe_data', null )` will be used instead:

```
const stripeServerData = 'undefined' !== typeof wc_stripe_upe_params
		         ? wc_stripe_upe_params
		         : wc.wcSettings.getSetting( 'stripe_data', null );
```

It's the same as in WC payments:

https://github.com/Automattic/woocommerce-payments/blob/a19aafbcbdb63e10921c7355b1858fd047969290/client/utils/checkout.js#L9-L17

File: client/stripe-utils/utils.js
Comment (diegocurbelo) [reply client/stripe-utils/utils.js:4]: I removed that `import` because internally it relies on the global `wc` variable, which is not defined when using the shortcode checkout (and now we are using the same `stripe-utils` files for both shortcode and blocks):

Screen Shot 2021-08-08 at 23 07 44

Is the `wc` global supposed to be defined for the shortcode checkout? maybe I'm missing a js enqueue?

File: includes/payment-methods/class-wc-stripe-upe-payment-gateway.php
Comment (diegocurbelo) [reply includes/payment-methods/class-wc-stripe-upe-payment-gateway.php]: Both the JS and the PHP were changed to use `key` instead of `publishableKey`.
The reason is that the current Stripe gateway uses `key` for both shortcode and blocks checkouts... the name `publishableKey ` was copied from WC Payments upe javascript.

File: includes/payment-methods/class-wc-stripe-upe-payment-gateway.php
Comment (diegocurbelo) [reply includes/payment-methods/class-wc-stripe-upe-payment-gateway.php:205]: The new UPE gateway does not extend the existing Stripe gateway (`WC_Gateway_Stripe`) as we won't be using any of the methods defined there... it extends directly from `WC_Stripe_Payment_Gateway`.

Right now we are using 4 of the 30+ configs that the other gateway has, if we need more then we can move the common config to a static helper since most of them are either nonces, order data, or calling other static methods.

File: includes/class-wc-stripe-intent-controller.php
Comment (diegocurbelo) [reply includes/class-wc-stripe-intent-controller.php:269]: The class `Process_Payment_Exception` does not exist in this repo, was a copy/paste mistake when copying the UPE code from the WC Payments repo (in the `feature/upe` PR)

In the plugin, we have only `WC_Stripe_Exception` but is just a wrapper with two messages (original and translated).
We shpuld do like in WC Payments, create a base exception class with a `message` and a string `error_code` and then create specific exceptions, in this case, `Process_Payment_Exception` with code `wc_stripe_upe_intent_error`.

But this PR is getting too big already, maybe we can add them in another PR?

File: client/stripe-utils/utils.js
Comment (reykjalin) [reply client/stripe-utils/utils.js]: Oof, I'm so sorry, you're right. I just read the code here wrong, nevermind this comment! 🤦‍♂️

File: client/stripe-utils/utils.js
Comment (reykjalin) [reply client/stripe-utils/utils.js:4]: > Is the `wc` global supposed to be defined for the shortcode checkout? maybe I'm missing a js enqueue? 

I thought it would be defined there as well, but you're right; it's not present. You're not missing a js enqueue 🙂 

Since that's the case, can we instead create a utility function that's a wrapper around us accessing the `wc` global? That way we can contain the global accessors to one file.

Something along the lines of:

```js
/* global wc_stripe_upe_params, wc */

export const getSettings = () => {
	return 'undefined' !== typeof wc_stripe_upe_params
		? wc_stripe_upe_params
		: wc.wcSettings.getSetting( 'stripe_data', null );
}
```

Comment (reykjalin) [reply client/stripe-utils/utils.js:4]: Also, we should probably use optional chaining to access the `wc` global:

```js
wc?.wcSettings?.getSetting( 'stripe_data', null );
```

File: includes/class-wc-stripe-intent-controller.php
Comment (reykjalin) [reply includes/class-wc-stripe-intent-controller.php:269]: Ahhh, sorry about that! I thought you were removing something that we had in Stripe already!

> this PR is getting too big already, maybe we can add them in another PR?

Yes, I agree. Let's just open a new issue for this since this is a more generic "use custom exceptions" issue than something specific to UPE or this PR 👍

File: includes/payment-methods/class-wc-stripe-upe-payment-gateway.php
Comment (reykjalin) [reply includes/payment-methods/class-wc-stripe-upe-payment-gateway.php:205]: > The new UPE gateway does not extend the existing Stripe gateway (`WC_Gateway_Stripe`) as we won't be using any of the methods defined there... it extends directly from `WC_Stripe_Payment_Gateway`.

Ah, you're right, I didn't realize the inheritance is from the abstract class 🤔 

This does bring up another thought for me though; maybe the `javascript_params()` function should be made part of the abstract `WC_Stripe_Payment_Gateway ` class 🤔. I think it'd certainly serve us well in this case, and it makes sense to have a re-usable function there.

What do you think?

I'm very concerned about repeating the configuration setup here because it's a massive configuration object. If we duplicate the functionality here it will eventually be a nightmare to maintain.

> Right now we are using 4 of the 30+ configs that the other gateway has, if we need more then we can move the common config to a static helper since most of them are either nonces, order data, or calling other static methods.

If this is the case, I wonder if it makes more sense to just create a separate configuration for the UPE class and remove the duplication from the Shortcode `javascript_params()` function?

If you're sure the configuration will be mostly distinct I'd be in favor of that. There's no need to construct a massive configuration object for just 4 properties.

To reduce the duplication we could probably just introduce methods to retrieve the values so we don't have to update them in multiple places if we need to change them in the future? I'm fine with skipping that if the configuration object for UPE is this small though, so I'll leave it to you to decide what makes the most sense here 🙂 

To reduce duplication this is what I'm thinking of:

```php
function javascript_params() {
	return [
		'option_a' => 'value_a',
		'option_b' => 'value_b',
	];
}
```

to

```php
function javascript_params() {
	return [
		'option_a' => $this->get_value_a(), // inherited function
		'option_b' => $this->get_value_b(), // inherited function
	];
}
```

File: includes/payment-methods/class-wc-stripe-upe-payment-gateway.php
Comment (reykjalin) [reply includes/payment-methods/class-wc-stripe-upe-payment-gateway.php]: Ah, sorry about that. I didn't realize this was aligning the copied WCPay code with pre-existing Stripe configuration. Ignore this comment then 🙂

File: client/classic/upe.js
Comment (reykjalin) [inline client/classic/upe.js]: I see ESLint complains about some style issues in this file. I think we should fix them, but some seem a bit unnecessary, e.g. using the short-hand syntax for functions instead of the long-form syntax in one case.

I thought I'd bring it up, but I'll leave it to @diegocurbelo to decide where and how to address this 🙂

File: client/classic/upe.js
Comment (reykjalin) [inline client/classic/upe.js]: I think this can be removed? We don't do the same type of fraud detection in Stripe that we do for WCPay 🤔

**Review by reykjalin — CHANGES_REQUESTED** [review]
Tests well, and the code looks good! ESLint reports some issues that I think we should fix before we merge though.

I'm not entirely sure what the purpose of `client/upe-styles` is so I'm not sure how thoroughly I should review that part of the code.
That said, the code looks good aside from the aforementioned issues reported by ESLint — I'm just not sure if the code is doing what it's supposed to be doing.

I'm also not sure how to run the unit tests 🤔 


---

If this specific PR is supposed to be tested by GlobalStep, let's be sure to add more specific testing instructions. If the entirety of UPE should be tested after a different PR has been merged, we don't need to change the instructions here.

We should, however, make sure there's a comprehensive list of tests we can provide to GlobalStep _somewhere_, and it'll be easier to maintain as part of a PR IMO.

---

I think it'd be clearer to move the UPE JS into `client/upe` and move `client/upe-styles` into `client/upe/styles`. When I see `client/classic` I sort of expect that to contain the "classic" shortcode implementation, i.e. what's now in `assets/js`. This may just be me though 🤔 

To be fair, the folder structure was already confusing, I just think it'd make more sense at the moment to use `client/upe` + `client/upe/styles`.

`client/classic/upe/index.js` and `client/classic/upe/styles` is also an option, but I feel like that's just creating a more nested folder structure for the sake of making a more nested folder structure 🙂 .

File: client/stripe-utils/utils.js
Comment (diegocurbelo) [reply client/stripe-utils/utils.js:4]: Added chaining to `wc` and `wcSettings()`

File: includes/payment-methods/class-wc-stripe-upe-payment-gateway.php
Comment (diegocurbelo) [reply includes/payment-methods/class-wc-stripe-upe-payment-gateway.php:205]: Comparing with what we have in WC Payments it doesn't seem like the gateways config have much in common (the same thing happens between `stripe` and `request-button` gateways), so I'm not so sure about having the `javascript_params()` as part of the abstract `WC_Stripe_Payment_Gateway` class, we would only have the `locale` and `ajax_url` there for now 😕... But this is definitely something we should do when/if the config grows.

File: client/classic/upe.js
Comment (diegocurbelo) [reply client/classic/upe.js]: You are right.. removed

File: client/classic/upe.js
Comment (diegocurbelo) [reply client/classic/upe.js]: Fixed

File: includes/payment-methods/class-wc-stripe-upe-payment-gateway.php
Comment (reykjalin) [reply includes/payment-methods/class-wc-stripe-upe-payment-gateway.php:205]: In that case let's just create a brand new `javascript_params()` function for the UPE checkout, with only the configurations necessary for the UPE checkout 🙂

**Comment by diegocurbelo** [comment]
> If this specific PR is supposed to be tested by GlobalStep, let's be sure to add more specific testing instructions. If the entirety of UPE should be tested after a different PR has been merged, we don't need to change the instructions here.

It's not supposed to be tested yet.. not until merging at least the happy flow (in PR #1659)

> We should, however, make sure there's a comprehensive list of tests we can provide to GlobalStep _somewhere_, and it'll be easier to maintain as part of a PR IMO.

Agree, and it should also be possible to copy them from WC Payments with minimum changes, as UPE should behave the s same for both.

> I think it'd be clearer to move the UPE JS into `client/upe` and move `client/upe-styles` into `client/upe/styles`. When I see `client/classic` I sort of expect that to contain the "classic" shortcode implementation, i.e. what's now in `assets/js`. This may just be me though 🤔

Good point... moved to `client/classic/upe` and client/styles/upe`

File: client/stripe-utils/utils.js
Comment (reykjalin) [reply client/stripe-utils/utils.js:4]: >Added chaining to `wc` and `wcSettings()`

🎉 

~~Can we add a wrapper function for retrieving the settings to limit the scope of where we access global values? Maybe this could be part of `getStripeServerData()`?~~

Nevermind, I'm a dumb-dumb. This is already _in_ `getStripeServerData()`. I can't facepalm hard enough 🤦‍♂️

**Review by reykjalin — APPROVED** [review]
Let's fix the linting errors reported by ESLint and then I think we're ready to merge 👏

**Pull request merged**
